// @medusajs/medusa/dist/api/middlewares/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformBody = exports.transformQuery = exports.canAccessBatchJob = exports.getRequestedBatchJob = void 0;
var authenticate_customer_1 = __importDefault(require("./authenticate-customer"));
var authenticate_1 = __importDefault(require("./authenticate"));
var normalized_query_1 = __importDefault(require("./normalized-query"));
var await_middleware_1 = __importDefault(require("./await-middleware"));
var get_requested_batch_job_1 = require("./batch-job/get-requested-batch-job");
Object.defineProperty(exports, "getRequestedBatchJob", { enumerable: true, get: function () { return get_requested_batch_job_1.getRequestedBatchJob; } });
var can_access_batch_job_1 = require("./batch-job/can-access-batch-job");
Object.defineProperty(exports, "canAccessBatchJob", { enumerable: true, get: function () { return can_access_batch_job_1.canAccessBatchJob; } });
var transform_query_1 = require("./transform-query");
Object.defineProperty(exports, "transformQuery", { enumerable: true, get: function () { return transform_query_1.transformQuery; } });
var transform_body_1 = require("./transform-body");
Object.defineProperty(exports, "transformBody", { enumerable: true, get: function () { return transform_body_1.transformBody; } });
exports.default = {
    authenticate: authenticate_1.default,
    authenticateCustomer: authenticate_customer_1.default,
    normalizeQuery: normalized_query_1.default,
    wrap: await_middleware_1.default,
};
