// @medusajs/medusa/dist/models/discount-condition.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiscountCondition = exports.DiscountConditionOperator = exports.DiscountConditionType = void 0;
var typeorm_1 = require("typeorm");
var db_aware_column_1 = require("../utils/db-aware-column");
var customer_group_1 = require("./customer-group");
var discount_rule_1 = require("./discount-rule");
var product_1 = require("./product");
var product_collection_1 = require("./product-collection");
var product_tag_1 = require("./product-tag");
var product_type_1 = require("./product-type");
var soft_deletable_entity_1 = require("../interfaces/models/soft-deletable-entity");
var generate_entity_id_1 = require("../utils/generate-entity-id");
var DiscountConditionType;
(function (DiscountConditionType) {
    DiscountConditionType["PRODUCTS"] = "products";
    DiscountConditionType["PRODUCT_TYPES"] = "product_types";
    DiscountConditionType["PRODUCT_COLLECTIONS"] = "product_collections";
    DiscountConditionType["PRODUCT_TAGS"] = "product_tags";
    DiscountConditionType["CUSTOMER_GROUPS"] = "customer_groups";
})(DiscountConditionType = exports.DiscountConditionType || (exports.DiscountConditionType = {}));
var DiscountConditionOperator;
(function (DiscountConditionOperator) {
    DiscountConditionOperator["IN"] = "in";
    DiscountConditionOperator["NOT_IN"] = "not_in";
})(DiscountConditionOperator = exports.DiscountConditionOperator || (exports.DiscountConditionOperator = {}));
var DiscountCondition = /** @class */ (function (_super) {
    __extends(DiscountCondition, _super);
    function DiscountCondition() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DiscountCondition.prototype.beforeInsert = function () {
        this.id = (0, generate_entity_id_1.generateEntityId)(this.id, "discon");
    };
    __decorate([
        (0, db_aware_column_1.DbAwareColumn)({
            type: "enum",
            enum: DiscountConditionType,
        }),
        __metadata("design:type", String)
    ], DiscountCondition.prototype, "type", void 0);
    __decorate([
        (0, db_aware_column_1.DbAwareColumn)({
            type: "enum",
            enum: DiscountConditionOperator,
        }),
        __metadata("design:type", String)
    ], DiscountCondition.prototype, "operator", void 0);
    __decorate([
        (0, typeorm_1.Index)(),
        (0, typeorm_1.Column)(),
        __metadata("design:type", String)
    ], DiscountCondition.prototype, "discount_rule_id", void 0);
    __decorate([
        (0, typeorm_1.ManyToOne)(function () { return discount_rule_1.DiscountRule; }, function (dr) { return dr.conditions; }),
        (0, typeorm_1.JoinColumn)({ name: "discount_rule_id" }),
        __metadata("design:type", discount_rule_1.DiscountRule)
    ], DiscountCondition.prototype, "discount_rule", void 0);
    __decorate([
        (0, typeorm_1.ManyToMany)(function () { return product_1.Product; }),
        (0, typeorm_1.JoinTable)({
            name: "discount_condition_product",
            joinColumn: {
                name: "condition_id",
                referencedColumnName: "id",
            },
            inverseJoinColumn: {
                name: "product_id",
                referencedColumnName: "id",
            },
        }),
        __metadata("design:type", Array)
    ], DiscountCondition.prototype, "products", void 0);
    __decorate([
        (0, typeorm_1.ManyToMany)(function () { return product_type_1.ProductType; }),
        (0, typeorm_1.JoinTable)({
            name: "discount_condition_product_type",
            joinColumn: {
                name: "condition_id",
                referencedColumnName: "id",
            },
            inverseJoinColumn: {
                name: "product_type_id",
                referencedColumnName: "id",
            },
        }),
        __metadata("design:type", Array)
    ], DiscountCondition.prototype, "product_types", void 0);
    __decorate([
        (0, typeorm_1.ManyToMany)(function () { return product_tag_1.ProductTag; }),
        (0, typeorm_1.JoinTable)({
            name: "discount_condition_product_tag",
            joinColumn: {
                name: "condition_id",
                referencedColumnName: "id",
            },
            inverseJoinColumn: {
                name: "product_tag_id",
                referencedColumnName: "id",
            },
        }),
        __metadata("design:type", Array)
    ], DiscountCondition.prototype, "product_tags", void 0);
    __decorate([
        (0, typeorm_1.ManyToMany)(function () { return product_collection_1.ProductCollection; }),
        (0, typeorm_1.JoinTable)({
            name: "discount_condition_product_collection",
            joinColumn: {
                name: "condition_id",
                referencedColumnName: "id",
            },
            inverseJoinColumn: {
                name: "product_collection_id",
                referencedColumnName: "id",
            },
        }),
        __metadata("design:type", Array)
    ], DiscountCondition.prototype, "product_collections", void 0);
    __decorate([
        (0, typeorm_1.ManyToMany)(function () { return customer_group_1.CustomerGroup; }),
        (0, typeorm_1.JoinTable)({
            name: "discount_condition_customer_group",
            joinColumn: {
                name: "condition_id",
                referencedColumnName: "id",
            },
            inverseJoinColumn: {
                name: "customer_group_id",
                referencedColumnName: "id",
            },
        }),
        __metadata("design:type", Array)
    ], DiscountCondition.prototype, "customer_groups", void 0);
    __decorate([
        (0, db_aware_column_1.DbAwareColumn)({ type: "jsonb", nullable: true }),
        __metadata("design:type", Object)
    ], DiscountCondition.prototype, "metadata", void 0);
    __decorate([
        (0, typeorm_1.BeforeInsert)(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], DiscountCondition.prototype, "beforeInsert", null);
    DiscountCondition = __decorate([
        (0, typeorm_1.Entity)(),
        (0, typeorm_1.Unique)("dctypeuniq", ["type", "operator", "discount_rule_id"])
    ], DiscountCondition);
    return DiscountCondition;
}(soft_deletable_entity_1.SoftDeletableEntity));
exports.DiscountCondition = DiscountCondition;
/**
 * @schema discount_condition
 * title: "Discount Condition"
 * description: "Holds rule conditions for when a discount is applicable"
 * x-resourceId: discount_condition
 * properties:
 *   id:
 *     description: "The id of the Discount Condition. Will be prefixed by `discon_`."
 *     type: string
 *   type:
 *     description: "The type of the Condition"
 *     type: string
 *     enum:
 *       - products
 *       - product_types
 *       - product_collections
 *       - product_tags
 *       - customer_groups
 *   created_at:
 *     description: "The date with timezone at which the resource was created."
 *     type: string
 *     format: date-time
 *   update_at:
 *     description: "The date with timezone at which the resource was last updated."
 *     type: string
 *     format: date-time
 *   deleted_at:
 *     description: "The date with timezone at which the resource was deleted."
 *     type: string
 *     format: date-time
 *   metadata:
 *     description: "An optional key-value map with additional information."
 *     type: object
 */
