// @medusajs/medusa-js/dist/resources/admin/batch-jobs.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var qs_1 = __importDefault(require("qs"));
var base_1 = __importDefault(require("../base"));
var utils_1 = require("../../utils");
var AdminBatchJobsResource = /** @class */ (function (_super) {
    __extends(AdminBatchJobsResource, _super);
    function AdminBatchJobsResource() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AdminBatchJobsResource.prototype.create = function (payload, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/batch-jobs";
        return this.client.request("POST", path, payload, {}, customHeaders);
    };
    AdminBatchJobsResource.prototype.list = function (query, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/batch-jobs";
        if (query) {
            var queryString = qs_1.default.stringify((0, utils_1.stringifyNullProperties)(query));
            path = "/admin/batch-jobs?" + queryString;
        }
        return this.client.request("GET", path, {}, {}, customHeaders);
    };
    AdminBatchJobsResource.prototype.cancel = function (batchJobId, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/batch-jobs/" + batchJobId + "/cancel";
        return this.client.request("POST", path, {}, {}, customHeaders);
    };
    AdminBatchJobsResource.prototype.confirm = function (batchJobId, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/batch-jobs/" + batchJobId + "/confirm";
        return this.client.request("POST", path, {}, {}, customHeaders);
    };
    AdminBatchJobsResource.prototype.retrieve = function (batchJobId, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/batch-jobs/" + batchJobId;
        return this.client.request("GET", path, {}, {}, customHeaders);
    };
    return AdminBatchJobsResource;
}(base_1.default));
exports.default = AdminBatchJobsResource;
