// @medusajs/medusa-js/dist/resources/admin/price-lists.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var qs_1 = __importDefault(require("qs"));
var base_1 = __importDefault(require("../base"));
var AdminPriceListResource = /** @class */ (function (_super) {
    __extends(AdminPriceListResource, _super);
    function AdminPriceListResource() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AdminPriceListResource.prototype.create = function (payload, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/price-lists";
        return this.client.request("POST", path, payload, {}, customHeaders);
    };
    AdminPriceListResource.prototype.update = function (id, payload, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/price-lists/" + id;
        return this.client.request("POST", path, payload, {}, customHeaders);
    };
    AdminPriceListResource.prototype.delete = function (id, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/price-lists/" + id;
        return this.client.request("DELETE", path, {}, {}, customHeaders);
    };
    AdminPriceListResource.prototype.retrieve = function (id, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/price-lists/" + id;
        return this.client.request("GET", path, {}, {}, customHeaders);
    };
    AdminPriceListResource.prototype.list = function (query, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/price-lists/";
        if (query) {
            var queryString = qs_1.default.stringify(query);
            path = "/admin/price-lists?" + queryString;
        }
        return this.client.request("GET", path, {}, {}, customHeaders);
    };
    AdminPriceListResource.prototype.listProducts = function (id, query, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/price-lists/" + id + "/products";
        if (query) {
            var queryString = qs_1.default.stringify(query);
            path = "/admin/price-lists/" + id + "/products?" + queryString;
        }
        return this.client.request("GET", path, {}, {}, customHeaders);
    };
    AdminPriceListResource.prototype.addPrices = function (id, payload, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/price-lists/" + id + "/prices/batch";
        return this.client.request("POST", path, payload, {}, customHeaders);
    };
    AdminPriceListResource.prototype.deletePrices = function (id, payload, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/price-lists/" + id + "/prices/batch";
        return this.client.request("DELETE", path, payload, {}, customHeaders);
    };
    AdminPriceListResource.prototype.deleteProductPrices = function (priceListId, productId, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/price-lists/" + priceListId + "/products/" + productId + "/prices";
        return this.client.request("DELETE", path, {}, {}, customHeaders);
    };
    AdminPriceListResource.prototype.deleteVariantPrices = function (priceListId, variantId, customHeaders) {
        if (customHeaders === void 0) { customHeaders = {}; }
        var path = "/admin/price-lists/" + priceListId + "/variants/" + variantId + "/prices";
        return this.client.request("DELETE", path, {}, {}, customHeaders);
    };
    return AdminPriceListResource;
}(base_1.default));
exports.default = AdminPriceListResource;
