// @medusajs/medusa/dist/services/draft-order.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
var medusa_core_utils_1 = require("medusa-core-utils");
var typeorm_1 = require("typeorm");
var models_1 = require("../models");
var interfaces_1 = require("../interfaces");
var utils_1 = require("../utils");
/**
 * Handles draft orders
 * @implements {BaseService}
 */
var DraftOrderService = /** @class */ (function (_super) {
    __extends(DraftOrderService, _super);
    function DraftOrderService(_a) {
        var manager = _a.manager, draftOrderRepository = _a.draftOrderRepository, paymentRepository = _a.paymentRepository, orderRepository = _a.orderRepository, eventBusService = _a.eventBusService, cartService = _a.cartService, lineItemService = _a.lineItemService, productVariantService = _a.productVariantService, shippingOptionService = _a.shippingOptionService;
        var _this = _super.call(this, {
            manager: manager,
            draftOrderRepository: draftOrderRepository,
            paymentRepository: paymentRepository,
            orderRepository: orderRepository,
            eventBusService: eventBusService,
            cartService: cartService,
            lineItemService: lineItemService,
            productVariantService: productVariantService,
            shippingOptionService: shippingOptionService,
        }) || this;
        _this.manager_ = manager;
        _this.draftOrderRepository_ = draftOrderRepository;
        _this.paymentRepository_ = paymentRepository;
        _this.orderRepository_ = orderRepository;
        _this.lineItemService_ = lineItemService;
        _this.cartService_ = cartService;
        _this.productVariantService_ = productVariantService;
        _this.shippingOptionService_ = shippingOptionService;
        _this.eventBus_ = eventBusService;
        return _this;
    }
    /**
     * Retrieves a draft order with the given id.
     * @param id - id of the draft order to retrieve
     * @param config - query object for findOne
     * @return the draft order
     */
    DraftOrderService.prototype.retrieve = function (id, config) {
        if (config === void 0) { config = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (transactionManager) { return __awaiter(_this, void 0, void 0, function () {
                            var draftOrderRepo, query, draftOrder;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        draftOrderRepo = transactionManager.getCustomRepository(this.draftOrderRepository_);
                                        query = (0, utils_1.buildQuery)({ id: id }, config);
                                        return [4 /*yield*/, draftOrderRepo.findOne(query)];
                                    case 1:
                                        draftOrder = _a.sent();
                                        if (!draftOrder) {
                                            throw new medusa_core_utils_1.MedusaError(medusa_core_utils_1.MedusaError.Types.NOT_FOUND, "Draft order with ".concat(id, " was not found"));
                                        }
                                        return [2 /*return*/, draftOrder];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Retrieves a draft order based on its associated cart id
     * @param cartId - cart id that the draft orders's cart has
     * @param config - query object for findOne
     * @return the draft order
     */
    DraftOrderService.prototype.retrieveByCartId = function (cartId, config) {
        if (config === void 0) { config = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (transactionManager) { return __awaiter(_this, void 0, void 0, function () {
                            var draftOrderRepo, query, draftOrder;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        draftOrderRepo = transactionManager.getCustomRepository(this.draftOrderRepository_);
                                        query = (0, utils_1.buildQuery)({ cart_id: cartId }, config);
                                        return [4 /*yield*/, draftOrderRepo.findOne(query)];
                                    case 1:
                                        draftOrder = _a.sent();
                                        if (!draftOrder) {
                                            throw new medusa_core_utils_1.MedusaError(medusa_core_utils_1.MedusaError.Types.NOT_FOUND, "Draft order was not found");
                                        }
                                        return [2 /*return*/, draftOrder];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Deletes draft order idempotently.
     * @param {string} draftOrderId - id of draft order to delete
     * @return {Promise<DraftOrder | undefined>} empty promise
     */
    DraftOrderService.prototype.delete = function (draftOrderId) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (transactionManager) { return __awaiter(_this, void 0, void 0, function () {
                            var draftOrderRepo, draftOrder;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        draftOrderRepo = transactionManager.getCustomRepository(this.draftOrderRepository_);
                                        return [4 /*yield*/, draftOrderRepo.findOne({
                                                where: { id: draftOrderId },
                                            })];
                                    case 1:
                                        draftOrder = _a.sent();
                                        if (!draftOrder) {
                                            return [2 /*return*/];
                                        }
                                        return [4 /*yield*/, draftOrderRepo.remove(draftOrder)];
                                    case 2: return [2 /*return*/, _a.sent()];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Lists draft orders alongside the count
     * @param selector - query selector to filter draft orders
     * @param config - query config
     * @return draft orders
     */
    DraftOrderService.prototype.listAndCount = function (selector, config) {
        if (config === void 0) { config = {
            skip: 0,
            take: 50,
            order: { created_at: "DESC" },
        }; }
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (transactionManager) { return __awaiter(_this, void 0, void 0, function () {
                            var draftOrderRepository, q, restSelector, query, where_1;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        draftOrderRepository = transactionManager.getCustomRepository(this.draftOrderRepository_);
                                        q = selector.q, restSelector = __rest(selector, ["q"]);
                                        query = (0, utils_1.buildQuery)(restSelector, config);
                                        if (q) {
                                            where_1 = query.where;
                                            where_1 === null || where_1 === void 0 ? true : delete where_1.display_id;
                                            query.join = {
                                                alias: "draft_order",
                                                innerJoin: {
                                                    cart: "draft_order.cart",
                                                },
                                            };
                                            query.where = function (qb) {
                                                qb.where(where_1);
                                                qb.andWhere(new typeorm_1.Brackets(function (qb) {
                                                    qb.where("cart.email ILIKE :q", {
                                                        q: "%".concat(q, "%"),
                                                    }).orWhere("draft_order.display_id::TEXT ILIKE :displayId", {
                                                        displayId: "".concat(q),
                                                    });
                                                }));
                                            };
                                        }
                                        return [4 /*yield*/, draftOrderRepository.findAndCount(query)];
                                    case 1: return [2 /*return*/, _a.sent()];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Lists draft orders
     * @param selector - query object for find
     * @param config - configurable attributes for find
     * @return list of draft orders
     */
    DraftOrderService.prototype.list = function (selector, config) {
        if (config === void 0) { config = {
            skip: 0,
            take: 50,
            order: { created_at: "DESC" },
        }; }
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (transactionManager) { return __awaiter(_this, void 0, void 0, function () {
                            var draftOrderRepo, query;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        draftOrderRepo = transactionManager.getCustomRepository(this.draftOrderRepository_);
                                        query = (0, utils_1.buildQuery)(selector, config);
                                        return [4 /*yield*/, draftOrderRepo.find(query)];
                                    case 1: return [2 /*return*/, _a.sent()];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Creates a draft order.
     * @param data - data to create draft order from
     * @return the created draft order
     */
    DraftOrderService.prototype.create = function (data) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (transactionManager) { return __awaiter(_this, void 0, void 0, function () {
                            var draftOrderRepo, shipping_methods, no_notification_order, items, rawCart, discounts, discounts_1, discounts_1_1, code, e_1_1, createdCart, draftOrder, result, items_1, items_1_1, item, line, price, e_2_1, shipping_methods_1, shipping_methods_1_1, method, e_3_1;
                            var e_1, _a, e_2, _b, e_3, _c;
                            return __generator(this, function (_d) {
                                switch (_d.label) {
                                    case 0:
                                        draftOrderRepo = transactionManager.getCustomRepository(this.draftOrderRepository_);
                                        if (!data.region_id) {
                                            throw new medusa_core_utils_1.MedusaError(medusa_core_utils_1.MedusaError.Types.INVALID_DATA, "region_id is required to create a draft order");
                                        }
                                        if (!data.items || !data.items.length) {
                                            throw new medusa_core_utils_1.MedusaError(medusa_core_utils_1.MedusaError.Types.INVALID_DATA, "Items are required to create a draft order");
                                        }
                                        shipping_methods = data.shipping_methods, no_notification_order = data.no_notification_order, items = data.items, rawCart = __rest(data, ["shipping_methods", "no_notification_order", "items"]);
                                        if (!rawCart.discounts) return [3 /*break*/, 8];
                                        discounts = rawCart.discounts;
                                        rawCart.discounts = [];
                                        _d.label = 1;
                                    case 1:
                                        _d.trys.push([1, 6, 7, 8]);
                                        discounts_1 = __values(discounts), discounts_1_1 = discounts_1.next();
                                        _d.label = 2;
                                    case 2:
                                        if (!!discounts_1_1.done) return [3 /*break*/, 5];
                                        code = discounts_1_1.value.code;
                                        return [4 /*yield*/, this.cartService_
                                                .withTransaction(transactionManager)
                                                .applyDiscount(rawCart, code)];
                                    case 3:
                                        _d.sent();
                                        _d.label = 4;
                                    case 4:
                                        discounts_1_1 = discounts_1.next();
                                        return [3 /*break*/, 2];
                                    case 5: return [3 /*break*/, 8];
                                    case 6:
                                        e_1_1 = _d.sent();
                                        e_1 = { error: e_1_1 };
                                        return [3 /*break*/, 8];
                                    case 7:
                                        try {
                                            if (discounts_1_1 && !discounts_1_1.done && (_a = discounts_1.return)) _a.call(discounts_1);
                                        }
                                        finally { if (e_1) throw e_1.error; }
                                        return [7 /*endfinally*/];
                                    case 8: return [4 /*yield*/, this.cartService_
                                            .withTransaction(transactionManager)
                                            .create(__assign({ type: models_1.CartType.DRAFT_ORDER }, rawCart))];
                                    case 9:
                                        createdCart = _d.sent();
                                        draftOrder = draftOrderRepo.create({
                                            cart_id: createdCart.id,
                                            no_notification_order: no_notification_order,
                                        });
                                        return [4 /*yield*/, draftOrderRepo.save(draftOrder)];
                                    case 10:
                                        result = _d.sent();
                                        return [4 /*yield*/, this.eventBus_
                                                .withTransaction(transactionManager)
                                                .emit(DraftOrderService.Events.CREATED, {
                                                id: result.id,
                                            })];
                                    case 11:
                                        _d.sent();
                                        _d.label = 12;
                                    case 12:
                                        _d.trys.push([12, 20, 21, 22]);
                                        items_1 = __values(items), items_1_1 = items_1.next();
                                        _d.label = 13;
                                    case 13:
                                        if (!!items_1_1.done) return [3 /*break*/, 19];
                                        item = items_1_1.value;
                                        if (!item.variant_id) return [3 /*break*/, 16];
                                        return [4 /*yield*/, this.lineItemService_
                                                .withTransaction(transactionManager)
                                                .generate(item.variant_id, data.region_id, item.quantity, {
                                                metadata: (item === null || item === void 0 ? void 0 : item.metadata) || {},
                                                unit_price: item.unit_price,
                                                cart: createdCart,
                                            })];
                                    case 14:
                                        line = _d.sent();
                                        return [4 /*yield*/, this.lineItemService_
                                                .withTransaction(transactionManager)
                                                .create(__assign(__assign({}, line), { cart_id: createdCart.id }))];
                                    case 15:
                                        _d.sent();
                                        return [3 /*break*/, 18];
                                    case 16:
                                        price = void 0;
                                        if (typeof item.unit_price === "undefined" || item.unit_price < 0) {
                                            price = 0;
                                        }
                                        else {
                                            price = item.unit_price;
                                        }
                                        // custom line items can be added to a draft order
                                        return [4 /*yield*/, this.lineItemService_
                                                .withTransaction(transactionManager)
                                                .create({
                                                cart_id: createdCart.id,
                                                has_shipping: true,
                                                title: item.title || "Custom item",
                                                allow_discounts: false,
                                                unit_price: price,
                                                quantity: item.quantity,
                                            })];
                                    case 17:
                                        // custom line items can be added to a draft order
                                        _d.sent();
                                        _d.label = 18;
                                    case 18:
                                        items_1_1 = items_1.next();
                                        return [3 /*break*/, 13];
                                    case 19: return [3 /*break*/, 22];
                                    case 20:
                                        e_2_1 = _d.sent();
                                        e_2 = { error: e_2_1 };
                                        return [3 /*break*/, 22];
                                    case 21:
                                        try {
                                            if (items_1_1 && !items_1_1.done && (_b = items_1.return)) _b.call(items_1);
                                        }
                                        finally { if (e_2) throw e_2.error; }
                                        return [7 /*endfinally*/];
                                    case 22:
                                        _d.trys.push([22, 27, 28, 29]);
                                        shipping_methods_1 = __values(shipping_methods), shipping_methods_1_1 = shipping_methods_1.next();
                                        _d.label = 23;
                                    case 23:
                                        if (!!shipping_methods_1_1.done) return [3 /*break*/, 26];
                                        method = shipping_methods_1_1.value;
                                        return [4 /*yield*/, this.cartService_
                                                .withTransaction(transactionManager)
                                                .addShippingMethod(createdCart.id, method.option_id, method.data)];
                                    case 24:
                                        _d.sent();
                                        _d.label = 25;
                                    case 25:
                                        shipping_methods_1_1 = shipping_methods_1.next();
                                        return [3 /*break*/, 23];
                                    case 26: return [3 /*break*/, 29];
                                    case 27:
                                        e_3_1 = _d.sent();
                                        e_3 = { error: e_3_1 };
                                        return [3 /*break*/, 29];
                                    case 28:
                                        try {
                                            if (shipping_methods_1_1 && !shipping_methods_1_1.done && (_c = shipping_methods_1.return)) _c.call(shipping_methods_1);
                                        }
                                        finally { if (e_3) throw e_3.error; }
                                        return [7 /*endfinally*/];
                                    case 29: return [2 /*return*/, result];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Registers a draft order as completed, when an order has been completed.
     * @param draftOrderId - id of draft order to complete
     * @param orderId - id of order completed from draft order cart
     * @return the created order
     */
    DraftOrderService.prototype.registerCartCompletion = function (draftOrderId, orderId) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (transactionManager) { return __awaiter(_this, void 0, void 0, function () {
                            var draftOrderRepo;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        draftOrderRepo = transactionManager.getCustomRepository(this.draftOrderRepository_);
                                        return [4 /*yield*/, draftOrderRepo.update({
                                                id: draftOrderId,
                                            }, {
                                                status: models_1.DraftOrderStatus.COMPLETED,
                                                completed_at: new Date(),
                                                order_id: orderId,
                                            })];
                                    case 1: return [2 /*return*/, _a.sent()];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Updates a draft order with the given data
     * @param id - id of the draft order
     * @param data - values to update the order with
     * @return the updated draft order
     */
    DraftOrderService.prototype.update = function (id, data) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.atomicPhase_(function (transactionManager) { return __awaiter(_this, void 0, void 0, function () {
                            var draftOrderRepo, draftOrder, touched;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        draftOrderRepo = transactionManager.getCustomRepository(this.draftOrderRepository_);
                                        return [4 /*yield*/, this.retrieve(id)];
                                    case 1:
                                        draftOrder = _a.sent();
                                        if (draftOrder.status === models_1.DraftOrderStatus.COMPLETED) {
                                            throw new medusa_core_utils_1.MedusaError(medusa_core_utils_1.MedusaError.Types.NOT_ALLOWED, "Can't update a draft order which is complete");
                                        }
                                        touched = false;
                                        if (data.no_notification_order !== undefined) {
                                            touched = true;
                                            draftOrder.no_notification_order = data.no_notification_order;
                                        }
                                        if (!touched) return [3 /*break*/, 4];
                                        return [4 /*yield*/, draftOrderRepo.save(draftOrder)];
                                    case 2:
                                        _a.sent();
                                        return [4 /*yield*/, this.eventBus_
                                                .withTransaction(transactionManager)
                                                .emit(DraftOrderService.Events.UPDATED, {
                                                id: draftOrder.id,
                                            })];
                                    case 3:
                                        _a.sent();
                                        _a.label = 4;
                                    case 4: return [2 /*return*/, draftOrder];
                                }
                            });
                        }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    DraftOrderService.Events = {
        CREATED: "draft_order.created",
        UPDATED: "draft_order.updated",
    };
    return DraftOrderService;
}(interfaces_1.TransactionBaseService));
exports.default = DraftOrderService;
