// @medusajs/medusa/dist/services/event-bus.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var bull_1 = __importDefault(require("bull"));
var ioredis_1 = __importDefault(require("ioredis"));
var sleep_1 = require("../utils/sleep");
/**
 * Can keep track of multiple subscribers to different events and run the
 * subscribers when events happen. Events will run asynchronously.
 */
var EventBusService = /** @class */ (function () {
    function EventBusService(_a, config, singleton) {
        var _this = this;
        var manager = _a.manager, logger = _a.logger, stagedJobRepository = _a.stagedJobRepository, redisClient = _a.redisClient, redisSubscriber = _a.redisSubscriber;
        if (singleton === void 0) { singleton = true; }
        /**
         * Handles incoming jobs.
         * @param job The job object
         * @return resolves to the results of the subscriber calls.
         */
        this.worker_ = function (job) { return __awaiter(_this, void 0, void 0, function () {
            var _a, eventName, data, eventObservers, wildcardObservers, observers;
            var _this = this;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = job.data, eventName = _a.eventName, data = _a.data;
                        eventObservers = this.observers_.get(eventName) || [];
                        wildcardObservers = this.observers_.get("*") || [];
                        observers = eventObservers.concat(wildcardObservers);
                        this.logger_.info("Processing ".concat(eventName, " which has ").concat(eventObservers.length, " subscribers"));
                        return [4 /*yield*/, Promise.all(observers.map(function (subscriber) {
                                return subscriber(data, eventName).catch(function (err) {
                                    _this.logger_.warn("An error occurred while processing ".concat(eventName, ": ").concat(err));
                                    console.error(err);
                                    return err;
                                });
                            }))];
                    case 1: return [2 /*return*/, _b.sent()];
                }
            });
        }); };
        /**
         * Handles incoming jobs.
         * @param job The job object
         * @return resolves to the results of the subscriber calls.
         */
        this.cronWorker_ = function (job) { return __awaiter(_this, void 0, void 0, function () {
            var _a, eventName, data, observers;
            var _this = this;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = job.data, eventName = _a.eventName, data = _a.data;
                        observers = this.cronHandlers_.get(eventName) || [];
                        this.logger_.info("Processing cron job: ".concat(eventName));
                        return [4 /*yield*/, Promise.all(observers.map(function (subscriber) {
                                return subscriber(data, eventName).catch(function (err) {
                                    _this.logger_.warn("An error occured while processing ".concat(eventName, ": ").concat(err));
                                    return err;
                                });
                            }))];
                    case 1: return [2 /*return*/, _b.sent()];
                }
            });
        }); };
        var opts = {
            createClient: function (type) {
                switch (type) {
                    case "client":
                        return redisClient;
                    case "subscriber":
                        return redisSubscriber;
                    default:
                        if (config.projectConfig.redis_url) {
                            return new ioredis_1.default(config.projectConfig.redis_url);
                        }
                        return redisClient;
                }
            },
        };
        this.config_ = config;
        this.manager_ = manager;
        this.logger_ = logger;
        this.stagedJobRepository_ = stagedJobRepository;
        if (singleton) {
            this.observers_ = new Map();
            this.queue_ = new bull_1.default("".concat(this.constructor.name, ":queue"), opts);
            this.cronHandlers_ = new Map();
            this.redisClient_ = redisClient;
            this.redisSubscriber_ = redisSubscriber;
            this.cronQueue_ = new bull_1.default("cron-jobs:queue", opts);
            // Register our worker to handle emit calls
            this.queue_.process(this.worker_);
            // Register cron worker
            this.cronQueue_.process(this.cronWorker_);
            if (process.env.NODE_ENV !== "test") {
                this.startEnqueuer();
            }
        }
    }
    EventBusService.prototype.withTransaction = function (transactionManager) {
        if (!transactionManager) {
            return this;
        }
        var cloned = new EventBusService({
            manager: transactionManager,
            stagedJobRepository: this.stagedJobRepository_,
            logger: this.logger_,
            redisClient: this.redisClient_,
            redisSubscriber: this.redisSubscriber_,
        }, this.config_, false);
        cloned.transactionManager_ = transactionManager;
        cloned.queue_ = this.queue_;
        return cloned;
    };
    /**
     * Adds a function to a list of event subscribers.
     * @param event - the event that the subscriber will listen for.
     * @param subscriber - the function to be called when a certain event
     * happens. Subscribers must return a Promise.
     * @return this
     */
    EventBusService.prototype.subscribe = function (event, subscriber) {
        var _a;
        if (typeof subscriber !== "function") {
            throw new Error("Subscriber must be a function");
        }
        var observers = (_a = this.observers_.get(event)) !== null && _a !== void 0 ? _a : [];
        this.observers_.set(event, __spreadArray(__spreadArray([], __read(observers), false), [subscriber], false));
        return this;
    };
    /**
     * Adds a function to a list of event subscribers.
     * @param event - the event that the subscriber will listen for.
     * @param subscriber - the function to be called when a certain event
     * happens. Subscribers must return a Promise.
     * @return this
     */
    EventBusService.prototype.unsubscribe = function (event, subscriber) {
        var _a, _b, _c;
        if (typeof subscriber !== "function") {
            throw new Error("Subscriber must be a function");
        }
        if ((_a = this.observers_.get(event)) === null || _a === void 0 ? void 0 : _a.length) {
            var index = (_b = this.observers_.get(event)) === null || _b === void 0 ? void 0 : _b.indexOf(subscriber);
            if (index !== -1) {
                (_c = this.observers_.get(event)) === null || _c === void 0 ? void 0 : _c.splice(index, 1);
            }
        }
        return this;
    };
    /**
     * Adds a function to a list of event subscribers.
     * @param event - the event that the subscriber will listen for.
     * @param subscriber - the function to be called when a certain event
     * happens. Subscribers must return a Promise.
     * @return this
     */
    EventBusService.prototype.registerCronHandler_ = function (event, subscriber) {
        var _a;
        if (typeof subscriber !== "function") {
            throw new Error("Handler must be a function");
        }
        var cronHandlers = (_a = this.cronHandlers_.get(event)) !== null && _a !== void 0 ? _a : [];
        this.cronHandlers_.set(event, __spreadArray(__spreadArray([], __read(cronHandlers), false), [subscriber], false));
        return this;
    };
    /**
     * Calls all subscribers when an event occurs.
     * @param {string} eventName - the name of the event to be process.
     * @param data - the data to send to the subscriber.
     * @param options - options to add the job with
     * @return the job from our queue
     */
    EventBusService.prototype.emit = function (eventName, data, options) {
        if (options === void 0) { options = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var stagedJobRepository, stagedJobInstance, opts;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.transactionManager_) return [3 /*break*/, 2];
                        stagedJobRepository = this.transactionManager_.getCustomRepository(this.stagedJobRepository_);
                        stagedJobInstance = stagedJobRepository.create({
                            event_name: eventName,
                            data: data,
                        });
                        return [4 /*yield*/, stagedJobRepository.save(stagedJobInstance)];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        opts = {
                            removeOnComplete: true,
                        };
                        if (typeof options.delay === "number") {
                            opts.delay = options.delay;
                        }
                        this.queue_.add({ eventName: eventName, data: data }, opts);
                        _a.label = 3;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    EventBusService.prototype.startEnqueuer = function () {
        this.shouldEnqueuerRun = true;
        this.enqueue_ = this.enqueuer_();
    };
    EventBusService.prototype.stopEnqueuer = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.shouldEnqueuerRun = false;
                        return [4 /*yield*/, this.enqueue_];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    EventBusService.prototype.enqueuer_ = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _loop_1, this_1;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _loop_1 = function () {
                            var listConfig, stagedJobRepo, jobs;
                            return __generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0:
                                        listConfig = {
                                            relations: [],
                                            skip: 0,
                                            take: 1000,
                                        };
                                        stagedJobRepo = this_1.manager_.getCustomRepository(this_1.stagedJobRepository_);
                                        return [4 /*yield*/, stagedJobRepo.find(listConfig)];
                                    case 1:
                                        jobs = _b.sent();
                                        return [4 /*yield*/, Promise.all(jobs.map(function (job) {
                                                _this.queue_
                                                    .add({ eventName: job.event_name, data: job.data }, { removeOnComplete: true })
                                                    .then(function () { return __awaiter(_this, void 0, void 0, function () {
                                                    return __generator(this, function (_a) {
                                                        switch (_a.label) {
                                                            case 0: return [4 /*yield*/, stagedJobRepo.remove(job)];
                                                            case 1:
                                                                _a.sent();
                                                                return [2 /*return*/];
                                                        }
                                                    });
                                                }); });
                                            }))];
                                    case 2:
                                        _b.sent();
                                        return [4 /*yield*/, (0, sleep_1.sleep)(3000)];
                                    case 3:
                                        _b.sent();
                                        return [2 /*return*/];
                                }
                            });
                        };
                        this_1 = this;
                        _a.label = 1;
                    case 1:
                        if (!this.shouldEnqueuerRun) return [3 /*break*/, 3];
                        return [5 /*yield**/, _loop_1()];
                    case 2:
                        _a.sent();
                        return [3 /*break*/, 1];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Registers a cron job.
     * @param eventName - the name of the event
     * @param data - the data to be sent with the event
     * @param cron - the cron pattern
     * @param handler - the handler to call on each cron job
     * @return void
     */
    EventBusService.prototype.createCronJob = function (eventName, data, cron, handler) {
        this.logger_.info("Registering ".concat(eventName));
        this.registerCronHandler_(eventName, handler);
        return this.cronQueue_.add({
            eventName: eventName,
            data: data,
        }, { repeat: { cron: cron } });
    };
    return EventBusService;
}());
exports.default = EventBusService;
